src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -213,24 +213,82 @@ public bool ExecuteReplicationOnce(TcpConnectionOptions tcpConnectionOptions, Ou
                     var skippedReplicationItemsInfo = new SkippedReplicationItemsInfo();
                     short lastTransactionMarker = -1;
                     long prevLastEtag = _lastEtag;
+                    var transactionMarkersToBeIncludedInBatch = new HashSet<short>();
 
                     using (_stats.Storage.Start())
-                    {                        
+                    {
                         foreach (var item in GetReplicationItems(documentsContext, _lastEtag, _stats))
                         {
                             _parent.CancellationToken.ThrowIfCancellationRequested();
+                            if (CanContinueBatch(ref next) == false)
+                            {
+                                wasInterrupted = true;
+                                break;
+                            }
 
-                            if (lastTransactionMarker != item.TransactionMarker)
+                            _stats.Storage.RecordInputAttempt();
+                            // here we add missing attachments in the same batch as the document that contains them without modifying the last etag or transaction boundary
+                            if (MissingAttachmentsInLastBatch &&
+                                item.Type == ReplicationBatchItem.ReplicationItemType.Document &&
+                                item is DocumentReplicationItem docItem &&
+                                docItem.Flags.Contain(DocumentFlags.HasAttachments))
                             {
+                                var type = (docItem.Flags & DocumentFlags.Revision) == DocumentFlags.Revision ? AttachmentType.Revision : AttachmentType.Document;
+                                foreach (var attachment in _parent._database.DocumentsStorage.AttachmentsStorage.GetAttachmentsForDocument(documentsContext, type, docItem.Id, docItem.ChangeVector))
+                                {
+                                    // we need to filter attachments that are been sent in the same batch as the document
+                                    if (attachment.Etag >= prevLastEtag)
+                                    {
+                                        Debug.Assert(docItem.TransactionMarker == attachment.TransactionMarker, $"{nameof(docItem)}.{nameof(docItem.TransactionMarker)} == {nameof(attachment)}.{nameof(attachment.TransactionMarker)}");
+                                        // treat the rare case when revision's attachment was created under different TransactionMarker
```

Comment:
Why is that a _valid_ case? When can that happen?

Reply:
We got it on few of our DBs I currently work on reproducing it in a test, but it seems to be OK to have different tx marker for document and its attachment, but should not happen for revision?

Reply:
A revision's attachment can be added by a previous revision, so that would be different.